File: service/app/gitea_client.py
"""Direct Gitea REST API calls (ADR-0004) — no `tea` CLI subprocess, no shell
interpolation of any model-derived string.
"""

from __future__ import annotations

import requests

from app.errors import GiteaError
from app.labels import DEFAULT_LABEL_COLOR, LABEL_COLORS
from app.schemas import GiteaIssue


class GiteaClient:
    def __init__(self, base_url: str, owner: str, repo: str, token: str, timeout: float = 15.0):
        self._base = f"{base_url}/api/v1/repos/{owner}/{repo}"
        self._session = requests.Session()
        self._session.headers.update({"Authorization": f"token {token}"})
        self._timeout = timeout
        self._label_id_cache: dict[str, int] | None = None

    def _request(self, method: str, path: str, **kwargs) -> requests.Response:
        try:
            resp = self._session.request(method, f"{self._base}{path}", timeout=self._timeout, **kwargs)
        except requests.RequestException as e:
            raise GiteaError(f"Gitea request failed: {e}") from e
        if resp.status_code >= 500:
            raise GiteaError(f"Gitea returned {resp.status_code}: {resp.text[:300]}")
        if resp.status_code >= 400:
            raise GiteaError(f"Gitea rejected request ({resp.status_code}): {resp.text[:300]}")
        return resp

    def _label_ids(self) -> dict[str, int]:
        if self._label_id_cache is None:
            resp = self._request("GET", "/labels")
            self._label_id_cache = {label["name"]: label["id"] for label in resp.json()}
        return self._label_id_cache

    def ensure_labels(self, names: list[str]) -> list[int]:
        """Create any of `names` that don't already exist in the repo, and
        return the full set's label ids. Idempotent — safe to call every
        startup/seed run.
        """
        ids = self._label_ids()
        missing = [n for n in names if n not in ids]
        for name in missing:
            resp = self._request(
                "POST",
                "/labels",
                json={"name": name, "color": LABEL_COLORS.get(name, DEFAULT_LABEL_COLOR)},
            )
            created = resp.json()
            ids[created["name"]] = created["id"]
        return [ids[n] for n in names]

    def create_issue(self, title: str, body: str, labels: list[str]) -> int:
        label_ids = self.ensure_labels(labels) if labels else []
        resp = self._request("POST", "/issues", json={"title": title, "body": body, "labels": label_ids})
        return resp.json()["number"]

    def comment_issue(self, issue_number: int, body: str) -> None:
        self._request("POST", f"/issues/{issue_number}/comments", json={"body": body})

    def list_open_issues(self) -> list[GiteaIssue]:
        issues: list[GiteaIssue] = []
        page = 1
        while True:
            resp = self._request(
                "GET",
                "/issues",
                params={"state": "open", "type": "issues", "page": page, "limit": 50},
            )
            batch = resp.json()
            if not batch:
                break
            issues.extend(_issue_from_json(raw) for raw in batch)
            if len(batch) < 50:
                break
            page += 1
        return issues

    def find_issue_by_title(self, title: str) -> GiteaIssue | None:
        resp = self._request("GET", "/issues", params={"state": "all", "q": title, "type": "issues"})
        for raw in resp.json():
            if raw["title"] == title:
                return _issue_from_json(raw)
        return None


def _issue_from_json(raw: dict) -> GiteaIssue:
    return GiteaIssue(
        number=raw["number"],
        title=raw["title"],
        body=raw.get("body") or "",
        labels=[label["name"] for label in raw.get("labels", [])],
        state=raw.get("state", "open"),
    )

File: service/src/gitea/gitea-client.ts
/**
 * Direct Gitea REST API calls (ADR-0004) — no `tea` CLI subprocess, no shell
 * interpolation of any model-derived string. Mirrors app/gitea_client.py.
 */

import { Inject, Injectable } from '@nestjs/common';
import { SETTINGS, Settings } from '../config/settings';
import { GiteaIssue } from '../reports/types';
import { GiteaError } from './gitea.errors';
import { DEFAULT_LABEL_COLOR, LABEL_COLORS } from './labels';

interface RawGiteaLabel {
  id: number;
  name: string;
}

interface RawGiteaIssue {
  number: number;
  title: string;
  body?: string | null;
  labels?: RawGiteaLabel[];
  state?: string;
}

const PAGE_SIZE = 50;
const REQUEST_TIMEOUT_MS = 15_000;

@Injectable()
export class GiteaClient {
  private readonly base: string;
  private readonly headers: Record<string, string>;
  private labelIdCache: Map<string, number> | null = null;

  constructor(@Inject(SETTINGS) settings: Settings) {
    this.base = `${settings.gitea_url}/api/v1/repos/${settings.gitea_repo_owner}/${settings.gitea_repo_name}`;
    this.headers = { Authorization: `token ${settings.gitea_token}`, 'Content-Type': 'application/json' };
  }

  private async request(method: string, path: string, body?: unknown): Promise<Response> {
    const response = await this.fetchResponse(method, path, body);
    await this.throwIfErrorStatus(response);
    return response;
  }

  private async fetchResponse(method: string, path: string, body?: unknown): Promise<Response> {
    try {
      return await fetch(`${this.base}${path}`, {
        method,
        headers: this.headers,
        body: body === undefined ? undefined : JSON.stringify(body),
        signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
      });
    } catch (err) {
      throw new GiteaError(`Gitea request failed: ${(err as Error).message}`);
    }
  }

  private async throwIfErrorStatus(response: Response): Promise<void> {
    if (response.status >= 500) {
      throw new GiteaError(`Gitea returned ${response.status}: ${(await response.text()).slice(0, 300)}`, response.status);
    }
    if (response.status >= 400) {
      throw new GiteaError(`Gitea rejected request (${response.status}): ${(await response.text()).slice(0, 300)}`, response.status);
    }
  }

  private async labelIds(): Promise<Map<string, number>> {
    if (this.labelIdCache === null) {
      const resp = await this.request('GET', '/labels');
      const raw = (await resp.json()) as RawGiteaLabel[];
      this.labelIdCache = new Map(raw.map((label) => [label.name, label.id]));
    }
    return this.labelIdCache;
  }

  /**
   * Create any of `names` that don't already exist in the repo, and return
   * the full set's label ids. Idempotent — safe to call every startup/seed
   * run.
   */
  async ensureLabels(names: readonly string[]): Promise<number[]> {
    const ids = await this.labelIds();
    const missing = names.filter((name) => !ids.has(name));
    for (const name of missing) {
      const resp = await this.request('POST', '/labels', {
        name,
        color: LABEL_COLORS[name] ?? DEFAULT_LABEL_COLOR,
      });
      const created = (await resp.json()) as RawGiteaLabel;
      ids.set(created.name, created.id);
    }
    return names.map((name) => {
      const id = ids.get(name);
      if (id === undefined) {
        throw new GiteaError(`Label "${name}" was neither found nor created`);
      }
      return id;
    });
  }

  async createIssue(title: string, body: string, labels: readonly string[]): Promise<number> {
    const labelIds = labels.length > 0 ? await this.ensureLabels(labels) : [];
    const resp = await this.request('POST', '/issues', { title, body, labels: labelIds });
    const created = (await resp.json()) as RawGiteaIssue;
    return created.number;
  }

  async commentIssue(issueNumber: number, body: string): Promise<void> {
    await this.request('POST', `/issues/${issueNumber}/comments`, { body });
  }

  async listOpenIssues(): Promise<GiteaIssue[]> {
    const issues: GiteaIssue[] = [];
    let page = 1;
    for (;;) {
      const resp = await this.request(
        'GET',
        `/issues?state=open&type=issues&page=${page}&limit=${PAGE_SIZE}`,
      );
      const batch = (await resp.json()) as RawGiteaIssue[];
      if (batch.length === 0) break;
      issues.push(...batch.map(fromRawIssue));
      if (batch.length < PAGE_SIZE) break;
      page += 1;
    }
    return issues;
  }

  async findIssueByTitle(title: string): Promise<GiteaIssue | null> {
    const resp = await this.request(
      'GET',
      `/issues?state=all&type=issues&q=${encodeURIComponent(title)}`,
    );
    const raw = (await resp.json()) as RawGiteaIssue[];
    const match = raw.find((issue) => issue.title === title);
    return match ? fromRawIssue(match) : null;
  }
}

function fromRawIssue(raw: RawGiteaIssue): GiteaIssue {
  return {
    number: raw.number,
    title: raw.title,
    body: raw.body ?? '',
    labels: (raw.labels ?? []).map((label) => label.name),
    state: raw.state ?? 'open',
  };
}
